### ISSUE TITLE
Two simple questions or maybe further improvements

### ISSUE THREAD

**Issue by JasonDoingGreat** [zylo117/Yet-Another-EfficientDet-Pytorch#126]
1. Since EfficientNet or the whole efficientdet consumes large GPU memory usage, sometimes it is impossible to use large batch size for some machines. Could it be helpful to replace BatchNorm with GroupNorm which appears more friendly to small batch size? I got the modified version of GroupNorm tested on my dataset and worked well, however, haven't tested on COCO.

2. In focal loss, simply append zero loss to empty-target data is not sufficient, if some pred boxes with high confidence appear, I think it would better to count the loss into total.

**Comment by zylo117** [comment]
1. I don't think effdet consumes large memory, comparing to yolo3 and rcnn.
2. can you specify the lines that you are talking about? can you share your modification?

**Comment by JasonDoingGreat** [comment]
1. Since I tried batch size of 4 or 8 on 2 cores of 1080Ti, it ends up out of gpu memory. (B4, train all layers) only batch size of 1 on each core could work. Simply parse and argument to choose between BatchNorm and GroupNorm is a good choice.
2. https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/blob/master/efficientdet/loss.py#L54
This part appears to me like append zero loss for images without any annotations. Sometimes in datasets there will be some data without annotations and use them as negative samples.

**Comment by zylo117** [comment]
1. yes, if you are using B4, then you are training D4, right? Then it takes 6G per image. But there is only 11G in 1080ti, so you can only do 1 batchsize per gpu. And in that case, syncbn can't even help. I guess GN may help, but without pretrained weights, you will have to train longer.

2. No the loss will be 0 even if you remove that part. 
check out here. https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/blob/master/efficientdet/loss.py#L66

If anchors can't match with annotations, like the ones that you called negative samples, the loss will 0.
But you don't need to worried about that, because for effdet, unlike rcnn, is a one-stage anchor type detection network, there is no need to add background class.

**Comment by JasonDoingGreat** [comment]
Thanks for the answers. Just tried GN and it run totally fine. Actually did not see much difference between BN and GN. might because of lack of data. usually how long does it take to train COCO with efficientdet-d4 to converge?

**Comment by zylo117** [comment]
Ages?
It depends on your devices, the original author train effdet on 32TPU, which is 1024G vram with 128 batchsize.
The loss will drop very quickly at first, of course, but then it will stagnate. I guess it's because effnet is a very deep CNN, not like resnet, B0 backbone has 270+ layers.

**Comment by JasonDoingGreat** [comment]
Thanks!

**Comment by zylo117** [comment]
If so, then you must have used a wrong argument, like trying to load d2 weights into a d0 network.
